File: docs/_code/consumer.cpp
#include <diaspora/Driver.hpp>
#include <diaspora/TopicHandle.hpp>
#include <iostream>

int main(int argc, char** argv) {

    if(argc != 3) {
        std::cerr << "Usage: "
                  << argv[0] << " <groupfile> <topic>" << std::endl;
        return -1;
    }

    auto group_file = argv[1];
    auto topic_name = argv[2];

    try {

        diaspora::Metadata options;
        options.json()["group_file"] = group_file;
        options.json()["margo"] = nlohmann::json::object();
        options.json()["margo"]["use_progress_thread"] = true;

        diaspora::Driver driver = diaspora::Driver::New("mofka", options);

        diaspora::TopicHandle topic = driver.openTopic(topic_name);

        diaspora::Consumer consumer = topic.consumer("consumer");
        for(size_t i = 0; i < 100; ++i) {
            diaspora::Event event = consumer.pull().wait(-1).value();
            std::cout << event.id() << ": " << event.metadata().string() << std::endl;
            if((i+1) % 10 == 0) event.acknowledge();
        }

    } catch(const diaspora::Exception& ex) {
        std::cerr << ex.what() << std::endl;
    }
    return 0;
}

File: docs/_code/consumer.py
import sys
import json
import pymargo.core
from pymargo.core import Engine
from diaspora_stream.api import Driver


def consume(group_file: str, topic_name: str):
    options = {
        "group_file": group_file,
        "margo": {
            "use_progress_thread": True
        }
    }
    driver = Driver.new("mofka", options)
    topic = driver.open_topic(topic_name)
    consumer = topic.consumer(name="myconsumer")

    for i in range(0, 100):
        event = consumer.pull().wait()
        print(event.metadata)
        if (i+1) % 10:
            event.acknowledge()


if __name__ == "__main__":

    if len(sys.argv) != 4:
        print(f"Usage: {sys.argv[0]} <groupfile> <topic>")
    group_file = sys.argv[1]
    topic_name = sys.argv[2]
    consume(group_file, topic_name)
